src/ngraph/pass/constant_folding_non_zero.cpp

```diff
@@ -0,0 +1,198 @@
+//*****************************************************************************
+// Copyright 2017-2020 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "constant_folding.hpp"
+#include "ngraph/op/non_zero.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+namespace
+{
+    struct NonZeroElements
+    {
+        using Results_t = std::vector<std::vector<int64_t>>;
+
+        NonZeroElements(const Shape& input_shape)
+            : m_input_shape{input_shape}
+            , m_results{Results_t(input_shape.size())}
+        {
+        }
+
+        template <typename T>
+        const Results_t& find_indices(const T* values)
+        {
+            m_current_index = Shape(m_input_shape.size(), 0UL);
+            const auto values_count = shape_size(m_input_shape);
+
+            const T zero_value = T{0};
+            for (size_t i = 0; i < values_count; ++i)
+            {
+                if (values[i] != zero_value)
+                {
+                    add_to_results(m_current_index);
+                }
+
+                // don't generate the next index when the last value has been processed
+                if (i < values_count - 1)
+                {
+                    next_index();
+                }
+            }
+
+            return m_results;
+        }
+
+    private:
+        /// \brief Adds single dimensions of an index into the matching element of the results
+        void add_to_results(const Shape& index)
+        {
+            for (size_t dim = 0; dim < index.size(); ++dim)
+            {
+                m_results.at(dim).push_back(index[dim]);
+            }
+        }
+
+        // Generates an index pointing to the next element in the flattened tensor
+        // It behaves similar to flipping bits when incrementing a binary number
+        void next_index()
+        {
+            for (size_t dim = m_current_index.size() - 1; dim >= 0; --dim)
```

Comment:
lets insert an assert here that m_current_index.size() != 0

Reply:
I'd rather do it once after this line https://github.com/NervanaSystems/ngraph/pull/4431/files?file-filters%5B%5D=.cpp#diff-7079b532eda2d11cc6adcf43a0a6d972R38
We don't have to check it every time `next_index()` is called, do we?

Reply:
By `assert` I assume you mean `throw`.

Reply:
I've added an NGRAPH_CHECK in this line https://github.com/NervanaSystems/ngraph/pull/4431/files?file-filters%5B%5D=.cpp#diff-7079b532eda2d11cc6adcf43a0a6d972R35